packages/hardhat/contracts/strategy/CrabMigration.sol

```diff
@@ -0,0 +1,417 @@
+            uint256 crabV1ToWithdraw = oSqthToPay.wmul(crabV1Balance).wdiv(data.v1oSqthToPay);
+            crabV1.withdraw(crabV1ToWithdraw);
+
+            crabV1.flashWithdraw(crabV1.balanceOf(address(this)), data.withdrawMaxEthToPay);
+            require(address(this).balance >= _amount, "M7");
+
+            if (data.ethToFlashDeposit > 0) {
+                crabV2.flashDeposit{value: address(this).balance - _amount}(data.ethToFlashDeposit);
+            }
+
+            // Sent back the V2 tokens to the user
+            crabV2.transfer(_initiator, crabV2.balanceOf(address(this)));
+
+            // Sent back the excess ETH
+            if (address(this).balance > _amount) {
+                payable(_initiator).sendValue(address(this).balance - _amount);
+            }
+        }
+    }
+
+    /**
+     * @notice allows users to claim their amount of crab v2 shares
+     */
+    function claimV2Shares() external afterMigration {
+        uint256 amountV1Deposited = sharesDeposited[msg.sender];
+        sharesDeposited[msg.sender] = 0;
+        uint256 amountV2ToTransfer = amountV1Deposited.wmul(totalCrabV2SharesReceived).wdiv(totalCrabV1SharesMigrated);
+        crabV2.transfer(msg.sender, amountV2ToTransfer);
+    }
+
+    /**
+     * @notice allows users to claim crabV2 shares and flash withdraw from crabV2
+     *
+     * @param _amountToWithdraw V2 shares to claim
+     * @param _maxEthToPay maximum ETH to pay to buy back the owed wSqueeth debt
+     */
+    function claimAndWithdraw(uint256 _amountToWithdraw, uint256 _maxEthToPay) external afterMigration {
+        uint256 amountV1toClaim = _getV1SharesForV2Share(_amountToWithdraw);
+        uint256 amountV1Deposited = sharesDeposited[msg.sender];
+        require(amountV1toClaim <= amountV1Deposited, "M6");
+
+        sharesDeposited[msg.sender] = amountV1Deposited.sub(amountV1toClaim);
+        crabV2.flashWithdraw(_amountToWithdraw, _maxEthToPay);
+
+        emit ClaimAndWithdraw(msg.sender, _amountToWithdraw);
+
+        // Pay user's ETH back
+        payable(msg.sender).sendValue(address(this).balance);
+    }
+
+    /**
+     * @notice For input v2 shares returns the equivalent v1 shares
+     */
+    function _getV1SharesForV2Share(uint256 _amountV2) internal view returns (uint256) {
+        return _amountV2.wmul(totalCrabV1SharesMigrated).wdiv(totalCrabV2SharesReceived);
+    }
+
+    function flashMigrationDetails(address _user)
+        external
+        view
+        returns (
+            bool,
+            uint256,
+            uint256,
+            uint256
+        )
+    {
+        return _flashMigrationDetails(_user);
+    }
+
+    /**
+     * @notice Used to migrate from crab V1 to crab V2 when CR1 >= CR2
+     *
+     * @param _ethToFlashDeposit Flash deposit amount in crab v2 with excess ETH. If 0 will returned to sender
+     */
+    function flashMigrateFromV1toV2(uint256 _ethToFlashDeposit) external afterMigration {
```

Comment:
It might be worth allowing partial positions to be flash migrated (in either function), just wondering if a whale doesnt do the batch and is stuck because migrating 100% of position would cause a lot of slippage.  Could be done by allowing someone to specify the amount of crab v1 to migrate vs balanceOf(msg.sender).